### ISSUE TITLE
RepositoriesService.ReplaceAllTopics has *Topics parameter, it'd be better made a value.

### ISSUE THREAD

**Issue by shurcooL** [google/go-github#773]
I just noticed the signature of [`RepositoriesService.ReplaceAllTopics`](https://godoc.org/github.com/google/go-github/github#RepositoriesService.ReplaceAllTopics) has a `*Topic` parameter.

The parameter is not optional, it's mandatory. See https://developer.github.com/v3/repos/#replace-all-topics-for-a-repository:

![image](https://user-images.githubusercontent.com/1924134/32302727-9388673a-bf3a-11e7-8caa-771a70b31af0.png)

There's no value in making it a pointer rather than value, only harm, because it makes it possible to pass `nil` value.

I suggest we making a breaking API change to improve the API of `go-github` and change the signature of the method to:

```Go
func (s *RepositoriesService) ReplaceAllTopics(ctx context.Context, owner, repo string,
    topics Topics) (*Topics, *Response, error)
```

This is a new method, still under a preview API, so a breaking change is acceptable IMO, as long as it leads to an improve `go-github` API.

/cc @gmlewis

**Comment by shurcooL** [comment]
Actually, better yet, we can also consider making it a `topics []string` instead:

```Go
func (s *RepositoriesService) ReplaceAllTopics(ctx context.Context, owner, repo string,
    topics []string) (*Topics, *Response, error)
```

**Comment by shurcooL** [comment]
If we go with `topics []string`, then #771 becomes no longer neccessary, since `Topics` won't necessarily be sent. It might instead be:

```Go
if topics == nil {
	topics = []string{}
}
t := struct {
	Names []string `json:"names"`
}{Names: topics}
resp, err := s.client.Do(ctx, req, t)
```

**Comment by jpbelangerupgrade** [comment]
I do like the simplicity of passing `[]string` instead of the `*Topics` struct which in the end only contains the string array. Also since this is a preview API, who knows what changes could be made by github on this, more breaking change could be coming anyway.

**Comment by jpbelanger-mtl** [comment]
Just let me know what you decide. Breaking API change is not a deal breaker on this, still a very simple preview feature from github.

**Comment by gmlewis** [comment]
Yes, I very much like changing the API to use `topics []string` in the parameter list, as [suggested by @shurcooL above](https://github.com/google/go-github/issues/773#issuecomment-341273316).

Thank you, @shurcooL and @jpbelanger-mtl!

**Comment by sahildua2305** [comment]
@shurcooL I think this issue can be closed. Is there anything left to be fixed after #771?

**Comment by shurcooL** [comment]
Thanks for noting that @sahildua2305, you're right.

Closing, this is resolved via #771.
